[BUG] 清空某个Pod annotations导致kruise-controller-manager崩溃重启

**问题描述**：
清空某个使用clone或asts（其他controller没有测试）运行的Pod的annotations时（Pod annotations字段变为空），kruise-controller-manager Pod 崩溃重启。
![image](https://github.com/openkruise/kruise/assets/91899187/ba3ca4f2-15d1-4f99-aa86-78193b09e5d6)

<details>
<summary>kruise-controller-manager 日志：</summary>
E0523 08:32:36.006302       1 runtime.go:78] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
goroutine 1227 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1b66920?, 0x214d810})
	/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x40d465?})
	/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x75
panic({0x1b66920, 0x214d810})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/openkruise/kruise/pkg/util/inplaceupdate.defaultPatchUpdateSpecToPod(0xc002ebec00, 0xc002d77cc0, 0xc002e8ea80)
	/workspace/pkg/util/inplaceupdate/inplace_update_defaults.go:173 +0x1169
github.com/openkruise/kruise/pkg/util/inplaceupdate.(*realControl).finishGracePeriod.func1()
	/workspace/pkg/util/inplaceupdate/inplace_update.go:220 +0x1ac
k8s.io/client-go/util/retry.OnError.func1()
	/workspace/vendor/k8s.io/client-go/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x1bf3240, 0x1})
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:217 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x216bc00?, 0xc000130008?}, 0x30?)
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:230 +0x57
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x1c8a020?)
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:223 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x40d7c7?)
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:418 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x1f6e8e0, 0xc00296d4a0)
	/workspace/vendor/k8s.io/client-go/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	/workspace/vendor/k8s.io/client-go/util/retry/util.go:104
github.com/openkruise/kruise/pkg/util/inplaceupdate.(*realControl).finishGracePeriod(0xc0007d0960?, 0xc002ec23e8?, 0xc002d77b80?)
	/workspace/pkg/util/inplaceupdate/inplace_update.go:186 +0x119
github.com/openkruise/kruise/pkg/util/inplaceupdate.(*realControl).Refresh(0xc00078da90?, 0xc002ec23e8, 0x1e30d66?)
	/workspace/pkg/util/inplaceupdate/inplace_update.go:114 +0x148
github.com/openkruise/kruise/pkg/controller/cloneset/sync.(*realControl).refreshPodState(0xc0003a33e0, 0xc002ef6b00, {0x2175540?, 0xc00078da90?}, 0xc002ec23e8, {0xc003874060, 0x10})
	/workspace/pkg/controller/cloneset/sync/cloneset_update.go:162 +0x15e
github.com/openkruise/kruise/pkg/controller/cloneset/sync.(*realControl).Update(0xc0003a33e0, 0xc002ef6b00, 0xc002ec6000, 0xc002ec6148, {0xc001887c50, 0x2, 0x2}, {0xc001887c30, 0x2, 0x2}, ...)
	/workspace/pkg/controller/cloneset/sync/cloneset_update.go:57 +0x15b
github.com/openkruise/kruise/pkg/controller/cloneset.(*ReconcileCloneSet).syncCloneSet(0xc001c2a1c0, 0xc001c40680, 0xc002e8e700, 0xc002ec6000, 0xc002ec6148, {0xc001887c50, 0x2, 0x2}, {0xc001887c30, 0x2, ...}, ...)
	/workspace/pkg/controller/cloneset/cloneset_controller.go:394 +0x41e
github.com/openkruise/kruise/pkg/controller/cloneset.(*ReconcileCloneSet).doReconcile(0xc001c2a1c0, {{{0xc00338bae0?, 0x1?}, {0xc00338bad0?, 0x4fe0e6?}}})
	/workspace/pkg/controller/cloneset/cloneset_controller.go:336 +0x2125
github.com/openkruise/kruise/pkg/controller/cloneset.(*ReconcileCloneSet).Reconcile(0x216bc70?, {0xc002d1aea0?, 0x1b63c80?}, {{{0xc00338bae0?, 0x1d6cec0?}, {0xc00338bad0?, 0x30b3ec0?}}})
	/workspace/pkg/controller/cloneset/cloneset_controller.go:180 +0x34
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc001cb8460, {0x216bc70, 0xc002d1aea0}, {{{0xc00338bae0?, 0x1d6cec0?}, {0xc00338bad0?, 0xc0014d5640?}}})
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x222
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc001cb8460, {0x216bbc8, 0xc0008fd400}, {0x1c06900?, 0xc0028250e0?})
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x2e9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc001cb8460, {0x216bbc8, 0xc0008fd400})
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x309
I0523 08:32:36.006392       1 cloneset_controller.go:190] Finished syncing CloneSet default/myapp, cost 5.790553ms
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 1227 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x40d465?})
	/workspace/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x1b66920, 0x214d810})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/openkruise/kruise/pkg/util/inplaceupdate.defaultPatchUpdateSpecToPod(0xc002ebec00, 0xc002d77cc0, 0xc002e8ea80)
	/workspace/pkg/util/inplaceupdate/inplace_update_defaults.go:173 +0x1169
github.com/openkruise/kruise/pkg/util/inplaceupdate.(*realControl).finishGracePeriod.func1()
	/workspace/pkg/util/inplaceupdate/inplace_update.go:220 +0x1ac
k8s.io/client-go/util/retry.OnError.func1()
	/workspace/vendor/k8s.io/client-go/util/retry/util.go:51 +0x33
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x1bf3240, 0x1})
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:217 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x216bc00?, 0xc000130008?}, 0x30?)
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:230 +0x57
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x1c8a020?)
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:223 +0x39
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x40d7c7?)
	/workspace/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:418 +0x5f
k8s.io/client-go/util/retry.OnError({0x989680, 0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0x1f6e8e0, 0xc00296d4a0)
	/workspace/vendor/k8s.io/client-go/util/retry/util.go:50 +0xf1
k8s.io/client-go/util/retry.RetryOnConflict(...)
	/workspace/vendor/k8s.io/client-go/util/retry/util.go:104
github.com/openkruise/kruise/pkg/util/inplaceupdate.(*realControl).finishGracePeriod(0xc0007d0960?, 0xc002ec23e8?, 0xc002d77b80?)
	/workspace/pkg/util/inplaceupdate/inplace_update.go:186 +0x119
github.com/openkruise/kruise/pkg/util/inplaceupdate.(*realControl).Refresh(0xc00078da90?, 0xc002ec23e8, 0x1e30d66?)
	/workspace/pkg/util/inplaceupdate/inplace_update.go:114 +0x148
github.com/openkruise/kruise/pkg/controller/cloneset/sync.(*realControl).refreshPodState(0xc0003a33e0, 0xc002ef6b00, {0x2175540?, 0xc00078da90?}, 0xc002ec23e8, {0xc003874060, 0x10})
	/workspace/pkg/controller/cloneset/sync/cloneset_update.go:162 +0x15e
github.com/openkruise/kruise/pkg/controller/cloneset/sync.(*realControl).Update(0xc0003a33e0, 0xc002ef6b00, 0xc002ec6000, 0xc002ec6148, {0xc001887c50, 0x2, 0x2}, {0xc001887c30, 0x2, 0x2}, ...)
	/workspace/pkg/controller/cloneset/sync/cloneset_update.go:57 +0x15b
github.com/openkruise/kruise/pkg/controller/cloneset.(*ReconcileCloneSet).syncCloneSet(0xc001c2a1c0, 0xc001c40680, 0xc002e8e700, 0xc002ec6000, 0xc002ec6148, {0xc001887c50, 0x2, 0x2}, {0xc001887c30, 0x2, ...}, ...)
	/workspace/pkg/controller/cloneset/cloneset_controller.go:394 +0x41e
github.com/openkruise/kruise/pkg/controller/cloneset.(*ReconcileCloneSet).doReconcile(0xc001c2a1c0, {{{0xc00338bae0?, 0x1?}, {0xc00338bad0?, 0x4fe0e6?}}})
	/workspace/pkg/controller/cloneset/cloneset_controller.go:336 +0x2125
github.com/openkruise/kruise/pkg/controller/cloneset.(*ReconcileCloneSet).Reconcile(0x216bc70?, {0xc002d1aea0?, 0x1b63c80?}, {{{0xc00338bae0?, 0x1d6cec0?}, {0xc00338bad0?, 0x30b3ec0?}}})
	/workspace/pkg/controller/cloneset/cloneset_controller.go:180 +0x34
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc001cb8460, {0x216bc70, 0xc002d1aea0}, {{{0xc00338bae0?, 0x1d6cec0?}, {0xc00338bad0?, 0xc0014d5640?}}})
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x222
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc001cb8460, {0x216bbc8, 0xc0008fd400}, {0x1c06900?, 0xc0028250e0?})
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x2e9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc001cb8460, {0x216bbc8, 0xc0008fd400})
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x309
</details>

**问题复现**：
```
apiVersion: apps.kruise.io/v1alpha1
kind: CloneSet
metadata:
  annotations:
    mydomain.com/app: myapp
  labels:
    app: myapp
    generated-by: kustomize
  name: myapp
  namespace: default
spec:
  replicas: 2
  selector:
    matchLabels:
      app: myapp
      generated-by: kustomize
  template:
    metadata:
      labels:
        app: myapp
        generated-by: kustomize
      annotations:                   # 运行后删除该行再次 apply
        mydomain.com/app: myapp      # 运行后删除该行再次 apply
    spec:
      containers:
      - name: myapp
        image: registry.mydomain.com/test/myapp:v1
        imagePullPolicy: IfNotPresent
  updateStrategy:
    inPlaceUpdateStrategy:
      gracePeriodSeconds: 10
    type: InPlaceIfPossible
```

**环境**：
- Kruise 版本: 1.4.0
- Kubernetes 版本: v1.20.15

--- a/pkg/util/inplaceupdate/inplace_update_defaults.go
+++ b/pkg/util/inplaceupdate/inplace_update_defaults.go
@@ -66,12 +66,7 @@ func SetOptionsDefaults(opts *UpdateOptions) *UpdateOptions {
 func defaultPatchUpdateSpecToPod(pod *v1.Pod, spec *UpdateSpec, state *appspub.InPlaceUpdateState) (*v1.Pod, error) {
 
 	klog.V(5).Infof("Begin to in-place update pod %s/%s with update spec %v, state %v", pod.Namespace, pod.Name, util.DumpJSON(spec), util.DumpJSON(state))
-	if pod.Labels == nil {
-		pod.Labels = make(map[string]string)
-	}
-	if pod.Annotations == nil {
-		pod.Annotations = make(map[string]string)
-	}
+
 	state.NextContainerImages = make(map[string]string)
 	state.NextContainerRefMetadata = make(map[string]metav1.ObjectMeta)
 
@@ -85,6 +80,13 @@ func defaultPatchUpdateSpecToPod(pod *v1.Pod, spec *UpdateSpec, state *appspub.I
 		if err = json.Unmarshal(modified, pod); err != nil {
 			return nil, err
 		}
+	}
+
+	if pod.Labels == nil {
+		pod.Labels = make(map[string]string)
+	}
+	if pod.Annotations == nil {
+		pod.Annotations = make(map[string]string)
 	}
 
 	// prepare containers that should update this time and next time, according to their priorities

